File: Dynamic_obstacle_avoidance/IIFDS.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# Time: 2021-3-9
# Author: ZYunfei
# Name: IIFDS-DDPG
# File func: enviroment class
import numpy as np
from Dynamic_obstacle_avoidance.dynamic_obstacle_environment import obs_list
from Dynamic_obstacle_avoidance.Method import getReward

class IIFDS:
    """使用IIFDS类训练时每次必须reset"""
    def __init__(self):
        """基本参数："""
        self.V0 = 1
        self.threshold = 0.2
        self.stepSize = 0.1
        self.lam = 8           # 越大考虑障碍物速度越明显

        self.obsR  = 1.5
        self.start = np.array([0,2,5],dtype=float)
        self.goal = np.array([10,10,5.5],dtype=float)

        self.timelog = 0        # 时间，用来计算动态障碍的位置
        self.timeStep = 0.1

        self.xmax = 10 / 180 * np.pi  # 偏航角速度最大值  每个步长允许变化的角度
        self.gammax = 10 / 180 * np.pi  # 爬升角速度最大值  每个步长允许变化的角度
        self.maximumClimbingAngle = 100 / 180 * np.pi  # 最大爬升角
        self.maximumSubductionAngle = - 75 / 180 * np.pi  # 最大俯冲角

        self.vObs = None
        self.vObsNext = None

        self.path = np.array([[]]).reshape(-1,3)        # 保存动态球的运动轨迹

        self.env_num = len(obs_list)
        self.env = obs_list[0]

    def reset(self):
        self.timelog = 0                                         # 重置时间
        self.path = np.array([[]]).reshape(-1, 3)                # 清空障碍路径记录表
        self.env = obs_list[np.random.randint(0, self.env_num)]  # 随机一个训练环境

    def updateObs(self,if_test=False):
        """返回位置与速度。"""
        if if_test:
            """测试环境"""
            self.timelog, dic = obs_list[4](self.timelog, self.timeStep)
        else:
            """否则用reset时随机的一个环境"""
            self.timelog, dic = self.env(self.timelog, self.timeStep)
        self.vObs = dic['v']
        self.path = np.vstack((self.path, dic['obsCenter']))
        return dic

    def calDynamicState(self, uavPos, obsCenter):
        """强化学习模型获得的state。"""
        s1 = (obsCenter - uavPos)*(self.distanceCost(obsCenter,uavPos)-self.obsR)/self.distanceCost(obsCenter,uavPos)
        s2 = self.goal - uavPos
        s3 = self.vObs
        return np.append(s1,[s2,s3])

    def calRepulsiveMatrix(self, uavPos, obsCenter, obsR, row0):
        n = self.partialDerivativeSphere(obsCenter, uavPos, obsR)
        tempD = self.distanceCost(uavPos, obsCenter) - obsR
        row = row0 * np.exp(1-1/(self.distanceCost(uavPos,self.goal)*tempD))
        T = self.calculateT(obsCenter, uavPos, obsR)
        repulsiveMatrix = np.dot(-n,n.T) / T**(1/row) / np.dot(n.T,n)[0][0]
        return repulsiveMatrix

    def calTangentialMatrix(self, uavPos, obsCenter, obsR, theta, sigma0):
        n = self.partialDerivativeSphere(obsCenter, uavPos, obsR)
        T = self.calculateT(obsCenter, uavPos, obsR)
        partialX = (uavPos[0] - obsCenter[0]) * 2 / obsR ** 2
        partialY = (uavPos[1] - obsCenter[1]) * 2 / obsR ** 2
        partialZ = (uavPos[2] - obsCenter[2]) * 2 / obsR ** 2
        tk1 = np.array([partialY, -partialX, 0],dtype=float).reshape(-1,1)
        tk2 = np.array([partialX*partialZ, partialY*partialZ, -partialX**2-partialY**2],dtype=float).reshape(-1,1)
        originalPoint = np.array([np.cos(theta), np.sin(theta), 0]).reshape(1,-1)
        tk = self.trans(originalPoint, tk1.squeeze(), tk2.squeeze(), n.squeeze())
        tempD = self.distanceCost(uavPos, obsCenter) - obsR
        sigma = sigma0 * np.exp(1-1/(self.distanceCost(uavPos,self.goal)*tempD))
        tangentialMatrix = tk.dot(n.T) / T**(1/sigma) / self.calVecLen(tk.squeeze()) / self.calVecLen(n.squeeze())
        return tangentialMatrix

    def getqNext(self, uavPos, obsCenter, vObs, row0, sigma0, theta, qBefore):
        u = self.initField(uavPos, self.V0, self.goal)
        repulsiveMatrix = self.calRepulsiveMatrix(uavPos, obsCenter, self.obsR, row0)
        tangentialMatrix = self.calTangentialMatrix(uavPos, obsCenter, self.obsR, theta, sigma0)
        T = self.calculateT(obsCenter, uavPos, self.obsR)
        vp = np.exp(-T / self.lam) * vObs
        M = np.eye(3) + repulsiveMatrix + tangentialMatrix
        ubar = (M.dot(u - vp.reshape(-1, 1)).T + vp.reshape(1, -1)).squeeze()
        # 限制ubar的模长，避免进入障碍内部后轨迹突变
        if self.calVecLen(ubar) > 5:
            ubar = ubar/self.calVecLen(ubar)*5
        if qBefore[0] is None:
            uavNextPos = uavPos + ubar * self.stepSize
        else:
            uavNextPos = uavPos + ubar * self.stepSize
            _, _, _, _, qNext = self.kinematicConstrant(uavPos, qBefore, uavNextPos)
        return uavNextPos

    def kinematicConstrant(self, q, qBefore, qNext):
        """
        运动学约束函数 返回(上一时刻航迹角，上一时刻爬升角，约束后航迹角，约束后爬升角，约束后下一位置qNext)
        """
        # 计算qBefore到q航迹角x1,gam1
        qBefore2q = q - qBefore
        if qBefore2q[0] != 0 or qBefore2q[1] != 0:
            x1 = np.arcsin(np.abs(qBefore2q[1] / np.sqrt(qBefore2q[0] ** 2 + qBefore2q[1] ** 2)))  # 这里计算的角限定在了第一象限的角 0-pi/2
            gam1 = np.arcsin(qBefore2q[2] / np.sqrt(np.sum(qBefore2q ** 2)))
        else:
            return None, None, None, None, qNext
        # 计算q到qNext航迹角x2,gam2
        q2qNext = qNext - q
        x2 = np.arcsin(np.abs(q2qNext[1] / np.sqrt(q2qNext[0] ** 2 + q2qNext[1] ** 2)))  # 这里同理计算第一象限的角度
        gam2 = np.arcsin(q2qNext[2] / np.sqrt(np.sum(q2qNext ** 2)))

        # 根据不同象限计算矢量相对于x正半轴的角度 0-2 * pi
        if qBefore2q[0] > 0 and qBefore2q[1] > 0:
            x1 = x1
        if qBefore2q[0] < 0 and qBefore2q[1] > 0:
            x1 = np.pi - x1
        if qBefore2q[0] < 0 and qBefore2q[1] < 0:
            x1 = np.pi + x1
        if qBefore2q[0] > 0 and qBefore2q[1] < 0:
            x1 = 2 * np.pi - x1
        if qBefore2q[0] > 0 and qBefore2q[1] == 0:
            x1 = 0
        if qBefore2q[0] == 0 and qBefore2q[1] > 0:
            x1 = np.pi / 2
        if qBefore2q[0] < 0 and qBefore2q[1] == 0:
            x1 = np.pi
        if qBefore2q[0] == 0 and qBefore2q[1] < 0:
            x1 = np.pi * 3 / 2


        # 根据不同象限计算与x正半轴的角度
        if q2qNext[0] > 0 and q2qNext[1] > 0:
            x2 = x2
        if q2qNext[0] < 0 and q2qNext[1] > 0:
            x2 = np.pi - x2
        if q2qNext[0] < 0 and q2qNext[1] < 0:
            x2 = np.pi + x2
        if q2qNext[0] > 0 and q2qNext[1] < 0:
            x2 = 2 * np.pi - x2
        if q2qNext[0] > 0 and q2qNext[1] == 0:
            x2 = 0
        if q2qNext[0] == 0 and q2qNext[1] > 0:
            x2 = np.pi / 2
        if q2qNext[0] < 0 and q2qNext[1] == 0:
            x2 = np.pi
        if q2qNext[0] == 0 and q2qNext[1] < 0:
            x2 = np.pi * 3 / 2

        # 约束航迹角x   xres为约束后的航迹角
        deltax1x2 = self.angleVec(q2qNext[0:2], qBefore2q[0:2])  # 利用点乘除以模长乘积求xoy平面投影的夹角
        if deltax1x2 < self.xmax:
            xres = x2
        elif x1 - x2 > 0 and x1 - x2 < np.pi:  # 注意这几个逻辑
            xres = x1 - self.xmax
        elif x1 - x2 > 0 and x1 - x2 > np.pi:
            xres = x1 + self.xmax
        elif x1 - x2 < 0 and x2 - x1 < np.pi:
            xres = x1 + self.xmax
        else:
            xres = x1 - self.xmax

        # 约束爬升角gam   注意：爬升角只用讨论在-pi/2到pi/2区间，这恰好与arcsin的值域相同。  gamres为约束后的爬升角
        if np.abs(gam1 - gam2) <= self.gammax:
            gamres = gam2
        elif gam2 > gam1:
            gamres = gam1 + self.gammax
        else:
            gamres = gam1 - self.gammax
        if gamres > self.maximumClimbingAngle:
            gamres = self.maximumClimbingAngle
        if gamres < self.maximumSubductionAngle:
            gamres = self.maximumSubductionAngle

        # 计算约束过后下一个点qNext的坐标
        Rq2qNext = self.distanceCost(q, qNext)
        deltax = Rq2qNext * np.cos(gamres) * np.cos(xres)
        deltay = Rq2qNext * np.cos(gamres) * np.sin(xres)
        deltaz = Rq2qNext * np.sin(gamres)

        qNext = q + np.array([deltax, deltay, deltaz])
        return x1, gam1, xres, gamres, qNext

    def loop(self):
        uavPos = self.start
        row0 = 0.5
        theta = 0.5
        sigma0 = 0.5
        path = self.start.reshape(1,-1)
        qBefore = [None, None, None]
        reward = 0
        for i in range(500):
            dic = self.updateObs(if_test=True)
            vObs, obsCenter = dic['v'], dic['obsCenter']
            uavNextPos = self.getqNext(uavPos, obsCenter, vObs, row0, sigma0, theta, qBefore)
            reward += getReward(obsCenter, uavNextPos, uavPos, qBefore, self)
            qBefore = uavPos
            uavPos = uavNextPos
            if self.distanceCost(uavPos,self.goal)<self.threshold:
                path = np.vstack((path, self.goal))
                _ = iifds.updateObs(if_test=True)
                break
            path = np.vstack((path, uavPos))
        print('路径的长度为:%f'%self.calPathLen(path))
        print('奖励为:%f' % reward)
        np.savetxt('./data_csv/pathMatrix.csv', path, delimiter=',')
        self.save_data()

    @staticmethod
    def distanceCost(point1, point2):
        return np.sqrt(np.sum((point1 - point2) ** 2))

    def initField(self, pos, V0, goal):
        """计算初始流场，返回列向量。"""
        temp1 = pos[0] - goal[0]
        temp2 = pos[1] - goal[1]
        temp3 = pos[2] - goal[2]
        temp4 = self.distanceCost(pos,goal)
        return -np.array([temp1,temp2,temp3],dtype=float).reshape(-1,1)*V0/temp4

    @staticmethod
    def partialDerivativeSphere(obs, pos, r):
        """计算球障碍物方程偏导数，返回列向量。"""
        temp1 = pos[0] - obs[0]
        temp2 = pos[1] - obs[1]
        temp3 = pos[2] - obs[2]
        return np.array([temp1,temp2,temp3],dtype=float).reshape(-1,1)*2/r**2

    @staticmethod
    def calculateT(obs, pos, r):
        """计算T。"""
        temp1 = pos[0] - obs[0]
        temp2 = pos[1] - obs[1]
        temp3 = pos[2] - obs[2]
        return (temp1**2 + temp2**2 + temp3**2)/r**2

    def calPathLen(self, path):
        """计算一个轨迹的长度。"""
        num = path.shape[0]
        len = 0
        for i in range(num-1):
            len += self.distanceCost(path[i,:], path[i+1,:])
        return len

    def trans(self, originalPoint, xNew, yNew, zNew):
        """
        坐标变换后地球坐标下坐标
        newX, newY, newZ是新坐标下三个轴上的方向向量
        返回列向量
        """
        lenx = self.calVecLen(xNew)
        cosa1 = xNew[0] / lenx
        cosb1 = xNew[1] / lenx
        cosc1 = xNew[2] / lenx

        leny = self.calVecLen(yNew)
        cosa2 = yNew[0] / leny
        cosb2 = yNew[1] / leny
        cosc2 = yNew[2] / leny

        lenz = self.calVecLen(zNew)
        cosa3 = zNew[0] / lenz
        cosb3 = zNew[1] / lenz
        cosc3 = zNew[2] / lenz

        B = np.array([[cosa1, cosb1, cosc1],
                      [cosa2, cosb2, cosc2],
                      [cosa3, cosb3, cosc3]],dtype=float)

        invB = np.linalg.inv(B)
        return np.dot(invB, originalPoint.T)



    @staticmethod
    def calVecLen(vec):
        """计算向量模长。"""
        return np.sqrt(np.sum(vec**2))

    @staticmethod
    def angleVec(vec1, vec2):  # 计算两个向量之间的夹角
        temp = np.dot(vec1, vec2) / np.sqrt(np.sum(vec1 ** 2)) / np.sqrt(np.sum(vec2 ** 2))
        temp = np.clip(temp, -1, 1)  # 可能存在精度误差导致上一步的temp略大于1，因此clip
        theta = np.arccos(temp)
        return theta


    def save_data(self):
        np.savetxt('./data_csv/start.csv', self.start, delimiter=',')
        np.savetxt('./data_csv/goal.csv', self.goal, delimiter=',')
        np.savetxt('./data_csv/obs_r.csv', np.array([self.obsR]), delimiter=',')
        np.savetxt('./data_csv/obs_trace.csv', self.path, delimiter=',')





if __name__ == "__main__":
    iifds = IIFDS()
    # uavPos = np.array([1,2,3])
    # obsCenter = np.array([2,7,3])
    # obsR = 1
    # row0 = 1
    #
    # print(iifds.calTangentialMatrix(uavPos, obsCenter, obsR, 0.2, 0.5))
    iifds.loop()

File: IIFDS_and_IFDS/IIFDS.m
% IIFDS算法实现
clc; clear; close all;
%% 参考文献：Formation Obstacle Avoidance: A Fluid-Based Solution
%% 初始参数设置
V0 = 1; 
threshold = 0.2;
stepSize = 0.1;
%% demo1:静态单球形障碍
% start = [0,1,5];
% goal = [10,10,6];
% obsCenter = [5,5,5.5];
% obsR = 1;
% row0 = 0.5;   % 反应系数
% sigma0 = 0.5; % 切向反应系数
% theta = 0;    % 切向方向系数
% % 绘制环境
% scatter3(start(1),start(2),start(3),60,"cyan",'filled','o','MarkerEdgeColor','k');hold on
% scatter3(goal(1),goal(2),goal(3),60,"magenta",'filled',"o",'MarkerEdgeColor','k')
% text(start(1),start(2),start(3),'  起点');
% text(goal(1),goal(2),goal(3),'  终点');
% drawSphere(obsCenter,obsR);
% xlabel('x(m)'); ylabel('y(m)'); zlabel('z(m)');
% title('UAV航迹规划路径'); axis equal;
% % 开始
% pos = start;
% path = pos;
% bar = [];
% for kk = 1:6
%     row0 = row0 + 0.2;
%     sigma0 = sigma0 + 0.2;
%     theta = theta + 0.2;
%     pos = start;
%     path = pos;
%     while distanceCost(pos, goal) > threshold
%         u = initField(pos, V0, goal);
%         % 计算排斥矩阵
%         n = partialDerivativeSphere(obsCenter, pos, obsR);
%         tempD = distanceCost(pos, obsCenter) - obsR; % 到障碍物表面距离
%         row = row0 * exp(1-1/(distanceCost(pos, goal)*tempD));
%         T = calculateT(obsCenter, pos, obsR);
%         repulsiveMatrix = - n * n' / T^(1/row) / (n'*n);
%         %计算切向矩阵
%         partialX = (pos(1) - obsCenter(1))*2/obsR^2;
%         partialY = (pos(2) - obsCenter(2))*2/obsR^2;
%         partialZ = (pos(3) - obsCenter(3))*2/obsR^2;
%         tk1 = [partialY, -partialX, 0]';
%         tk2 = [partialX*partialZ, partialY*partialZ, -partialX^2-partialY^2]';
%         tk = trans([cos(theta), sin(theta), 0], tk1, tk2, n);
%         sigma = sigma0 * exp(1-1/(distanceCost(pos, goal)*tempD));
%         tangentialMatrix = tk * n' / T^(1/sigma) / calVecLen(tk) / calVecLen(n);
%         % 计算修正后扰动矩阵
%         M = eye(3) + repulsiveMatrix + tangentialMatrix;
%         ubar = (M * u)';
%         nextPos = pos + ubar * stepSize;
%         path = [path;nextPos];
%         pos = nextPos;
%     end
%     path = [path;goal];
%     b = plot3(path(:,1), path(:,2), path(:,3),'LineWidth',2); hold on;
%     bar = [bar,b];
% end
% legend(bar,["=","=","=","=","=","="]);
%% demo2:单个动态球障碍
start = [1,4,5];
goal = [10,10,5.5];
obsStart = [5,5,5]; % 动态障碍物开始的点位
obsR = 1.5;           % 动态障碍物的半径
lambda = 8;
row0 = 0.5;   % 反应系数
sigma0 = 0.5; % 切向反应系数
theta = 0.5;    % 切向方向系数
scatter3(start(1),start(2),start(3),60,"cyan",'filled','o','MarkerEdgeColor','k');hold on
scatter3(goal(1),goal(2),goal(3),60,"magenta",'filled',"o",'MarkerEdgeColor','k')
text(start(1),start(2),start(3),'  起点');
text(goal(1),goal(2),goal(3),'  终点');
xlabel('x(km)'); ylabel('y(km)'); zlabel('z(km)');
title('UAV动态航迹规划路径'); axis equal;
uavPos = start;
path = uavPos;
t = 0; % 初始化时间
timeStep = 0.1;
flag = true; % 是否通过碰撞检测
collPoint = [];
while distanceCost(uavPos, goal) > threshold
    obsCenter = [obsStart(1)+2*cos(t),obsStart(2)+2*sin(t),obsStart(3)];
    vObs = [-2*sin(t),2*cos(t),0]; % vObs是obsPos对t的导数
    try delete(B1), catch, end
    try delete(B2), catch, end
    B1 = drawSphere(obsCenter, obsR);
    B2 = scatter3(uavPos(1),uavPos(2),uavPos(3),80,'filled',"^",'MarkerFaceColor','g'...
                  ,'MarkerEdgeColor','k');
    b1 = plot3([obsStart(1)+2*cos(t-timeStep),obsCenter(1)],[obsStart(2)+2*sin(t-timeStep),obsCenter(2)]...
              ,[obsStart(3),obsStart(3)],'LineWidth',2,'color','b');
    drawnow;
    % 碰撞检测
    if distanceCost(uavPos, obsCenter) <= obsR
        flag = false;
        collPoint = [collPoint; uavPos];
    end
    % 算法开始
    u = initField(uavPos,V0,goal);
    % 计算排斥矩阵
    n = partialDerivativeSphere(obsCenter, uavPos, obsR);
    tempD = distanceCost(uavPos, obsCenter) - obsR; % 到障碍物表面距离
    row = row0 * exp(1-1/(distanceCost(uavPos, goal)*tempD));
    T = calculateT(obsCenter, uavPos, obsR);
    repulsiveMatrix = - n * n' / T^(1/row) / (n'*n);
    %计算切向矩阵
    partialX = (uavPos(1) - obsCenter(1))*2/obsR^2;
    partialY = (uavPos(2) - obsCenter(2))*2/obsR^2;
    partialZ = (uavPos(3) - obsCenter(3))*2/obsR^2;
    tk1 = [partialY, -partialX, 0]';
    tk2 = [partialX*partialZ, partialY*partialZ, -partialX^2-partialY^2]';
    tk = trans([cos(theta), sin(theta), 0], tk1, tk2, n);
    sigma = sigma0 * exp(1-1/(distanceCost(uavPos, goal)*tempD));
    tangentialMatrix = tk * n' / T^(1/sigma) / calVecLen(tk) / calVecLen(n);
    % 计算vp
    vp = exp(-T/lambda)*vObs; 
    % 计算修正后扰动矩阵
    M = eye(3) + repulsiveMatrix + tangentialMatrix;
    ubar = (M * (u-vp'))'+vp;
    uavNextPos = uavPos + ubar * stepSize;
    b2 = plot3([uavPos(1),uavNextPos(1)],[uavPos(2),uavNextPos(2)],[uavPos(3),uavNextPos(3)],'LineWidth',2,'color','r');
    path = [path;uavNextPos];
    uavPos = uavNextPos;
    t = t + timeStep; 
    pause(0.01)
    if t == 0 + timeStep
        legend([b1,b2,B2],["障碍物移动轨迹","UAV规划航路","UAV"],'AutoUpdate','off','Location','west')
    end
end
path = [path;goal];
plot3([path(end,1),goal(1)],[path(end,2),goal(2)],[path(end,3),goal(3)],'LineWidth',2,'color','r');
if flag == true
    fprintf('通过碰撞检测！')
else
    fprintf('未通过碰撞检测！\n')
    disp(collPoint);
end
%% 计算GS,LS,L
pathLength = 0;
for i=1:length(path(:,1))-1, pathLength = pathLength + distanceCost(path(i,1:3),path(i+1,1:3)); end
fprintf("航路长度为:%f\n GS:%f °\n LS:%f °",pathLength, calGs(path)/pi*180, calLs(path)/pi*180);
%% 函数
% 欧式距离求解函数
function h=distanceCost(a,b)
h = sqrt(sum((a-b).^2, 2));
end
% 初始流场求解函数
function u = initField(pos, C, goal)
u = -[(pos(1)-goal(1)), (pos(2)-goal(2)), (pos(3)-goal(3))]' * C / distanceCost(pos, goal);
end
% 球绘制函数
function bar = drawSphere(pos, r)
[x,y,z] = sphere(60);
bar = surf(r*x+pos(1), r*y+pos(2), r*z+pos(3));
hold on;
end
% 球障碍物方程偏导计算函数
function pd = partialDerivativeSphere(obs, pos, r)
pd = [pos(1)-obs(1), pos(2)-obs(2), pos(3)-obs(3)]'*2/r^2;
end
% 球障碍物计算T值
function T = calculateT(obs, pos, r)
T = ((pos(1)-obs(1))^2+(pos(2)-obs(2))^2+(pos(3)-obs(3))^2)/r^2;
end
% 坐标变换后地球坐标下坐标
% newX, newY, newZ是新坐标下三个轴上的方向向量
function pos = trans(originalPoint, xNew, yNew, zNew)
cosa1 = xNew(1)/sqrt(sum(xNew.^2));
cosb1 = xNew(2)/sqrt(sum(xNew.^2));
cosc1 = xNew(3)/sqrt(sum(xNew.^2));

cosa2 = yNew(1)/sqrt(sum(yNew.^2));
cosb2 = yNew(2)/sqrt(sum(yNew.^2));
cosc2 = yNew(3)/sqrt(sum(yNew.^2));

cosa3 = zNew(1)/sqrt(sum(zNew.^2));
cosb3 = zNew(2)/sqrt(sum(zNew.^2));
cosc3 = zNew(3)/sqrt(sum(zNew.^2));

B = [cosa1, cosb1, cosc1;
    cosa2, cosb2, cosc2;
    cosa3, cosb3, cosc3];
pos = B\originalPoint';
end
% 求向量模长
function L = calVecLen(vec)
L = sqrt(sum(vec.^2));
end






